### PR TITLE
fvalues

### PR DESCRIPTION
**Pull request by alexmojaki**
- Adds support for the new library https://github.com/oughtinc/fvalues which was made for this. In particular, `F` objects (strings) are serialised specially and then rendered specially in the frontend.
- Tweaks the primer debate recipe as an example of using `F`.
- Replaces `JSONEncoder` with `to_json_value` which recursively converts arbitrary objects to JSON-serialisable values. This is primarily for `F` objects: `default` doesn't work since `F` is a subclass of `str` which is already serialisable. But I also find it helpful in general to do these conversions before serialisation, e.g. `_get_first_descendant` was repeating logic.
- `to_json_value` also handles dataclass instances. Not particularly related to the rest of the PR.
- `compress` is replaced by `Paper.dict()`.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -14,9 +14,10 @@ classifiers =
 [options]
 packages = ice
 install_requires =
-    Faker
     defopt
+    faker
```
Comment (lslunis) [inline setup.cfg:18]: Is this case-insensitive?

File: ui/components/TracePage/FString.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+          <span
+            key={i}
+            title={part.source}
+            className={`p2 bg-${color}-50 text-${color}-600`}
+            // style={{
+            //   border: "1px solid #ccc",
+            //   borderRadius: "4px",
+            //   padding: "1px 2px",
+            //   display: "inline-block",
+            // }}
```
Comment (lslunis) [inline ui/components/TracePage/FString.tsx:55]: As above

File: ui/components/TracePage/FString.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        const inner = part.formatted;
+
+        // This commented code allows seeing the original nested construction,
+        // if the value isn't flattened.
+        // const inner =
+        //   typeof part.value === "object" && "__fstring__" in part.value ? (
+        //     <FString parts={part.value.__fstring__ as FStringPart[]} />
+        //   ) : (
+        //     part.formatted
+        //   );
```
Comment (lslunis) [inline ui/components/TracePage/FString.tsx:43]: I think we should either add this as an option or remove the comment. I lean toward removing it

File: ui/components/TracePage/FString.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        return (
+          <span
+            key={i}
+            title={part.source}
```
Comment (lslunis) [inline ui/components/TracePage/FString.tsx:48]: Let's use https://chakra-ui.com/docs/components/tooltip

**Review by lslunis — APPROVED** [review]

File: setup.cfg [hunk 5]
```diff
@@ -14,9 +14,10 @@ classifiers =
...
     fastapi
+    fvalues
```
Comment (lslunis) [inline setup.cfg:20]: Could you give more Oughters access to the fvalue repo/package so we can make changes if you're on vacation?

**Review by lslunis — APPROVED** [review]

File: ice/recipes/primer/debate/prompt.py [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
-    prompt = f"""
+    prompt = F(
+        f"""
 You are {agent_name}. There are {turns_left} turns left in the debate. You are trying to win the debate using reason and evidence. Don't repeat yourself. No more than 1-2 sentences per turn.
 
 {render_debate(debate, agent_name)}
 You: "
-""".strip()
+"""
+    ).strip()
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply ice/recipes/primer/debate/prompt.py:14]: I think we need to release this stuff on PyPI first.

File: ice/json_value.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+            k if isinstance(k, str) else repr(k): to_json_value(v) for k, v in x.items()
+        }
+    if isinstance(x, (list, tuple, set)):
+        return [to_json_value(v) for v in x]
+    if isinstance(x, F):
+        return {"__fstring__": to_json_value(x.parts)}
+    if hasattr(x, "dict") and callable(x.dict):
+        try:
+            return to_json_value(x.dict())
+        except TypeError:
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply ice/json_value.py:24]: I was just imitating the previous code, but done.

File: ice/recipes/primer/debate/prompt.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexmojaki) [reply ice/recipes/primer/debate/prompt.py:14]: https://linear.app/ought/issue/ICE-35/new-pypi-release

File: ice/recipes/primer/debate/prompt.py [hunk 8]
```diff
@@ -1,11 +1,15 @@
+from fvalues import F
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply ice/recipes/primer/debate/prompt.py:1]: https://linear.app/ought/issue/ICE-37/update-primer-with-new-features

File: setup.cfg (hunk 1 above)
Comment (alexmojaki) [reply setup.cfg:18]: Yes, pip and PyPI are case insensitive, and also -/_ are interchangeable.

File: ui/components/TracePage/FString.tsx (hunk 4 above)
Comment (alexmojaki) [reply ui/components/TracePage/FString.tsx]: Tried this now: https://github.com/oughtinc/ice/compare/f...f-tooltip

The CSS is being weird and I don't think it's worth blocking this to figure it out, having any kind of tooltip is already a bonus.

https://linear.app/ought/issue/ICE-38/use-chakra-tooltip-in-f-values

File: setup.cfg (hunk 5 above)
Comment (alexmojaki) [reply setup.cfg:20]: I'm not familiar with how this stuff works with organisations/teams, it seems pretty weird. I actually had to ask Andreas to make the repo public, it wouldn't let me. Anyway, I added you as an admin and the core team as maintainers.

In https://github.com/oughtinc/fvalues/settings/access it says:

> No base role set. All [Members](https://github.com/orgs/oughtinc/people) can access this repository.

I don't know what 'access' means, given that it's publicly visible.

**Review by lslunis — APPROVED** [review]

**Pull request merged**